File: camfit-puller/src/camfit_puller/ports/source.py
from __future__ import annotations
from typing import Iterator, Optional, Protocol, runtime_checkable
from ..domain.models import Camp, Review


@runtime_checkable
class DataSource(Protocol):
    name: str
    def iter_summaries(self) -> Iterator[Camp]: ...
    def get_detail(self, camp_id: str) -> Optional[Camp]: ...
    def iter_reviews(self, camp_id: str, *, sort: str = "recommend") -> Iterator[Review]: ...
    def iter_filters(self) -> Iterator[tuple[str, str, str, dict | None]]:
        """yield (id, name, kind, raw_json) for each native taxonomy entry."""
        ...

File: backend/be-api/internal/ports/source.go
package ports

import (
	"context"

	"github.com/whyjp/cf/be-api/internal/domain"
)

// SourceReader is the iteration-style data-source port. 1:1 with Python
// `ports.source.DataSource`, using channels instead of generators.
//
// The `errs` channel is single-buffered so the producer can post a final
// error without the consumer racing on close. The consumer should always
// drain `out` *and* check `errs` after the channel is closed.
type SourceReader interface {
	Name() string
	IterSummaries(ctx context.Context) (<-chan *domain.Camp, <-chan error)
	GetDetail(ctx context.Context, campID string) (*domain.Camp, error)
	IterReviews(ctx context.Context, campID string, sort string) (<-chan *domain.Review, <-chan error)
	IterFilters(ctx context.Context) (<-chan FilterEntry, <-chan error)
}

// FilterEntry mirrors the Python tuple `(id, name, kind, raw_json)`.
type FilterEntry struct {
	ID   string
	Name string
	Kind string
	Raw  map[string]any
}
